# Task 2

In [1]:
# import python packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
from sklearn import svm
from sklearn.feature_selection import SelectKBest,chi2
import seaborn as sns

In [2]:
from sklearn.datasets import fetch_rcv1

In [3]:
rcv1 = fetch_rcv1()

In [4]:
rcv1.target.shape

(804414, 103)

In [5]:
pd.DataFrame.sparse.from_spmatrix(rcv1.target)

0    1    2    3    4    5    6    7    8    9    ...  93   94   95   \
0         0    0    0    0    0    0    0    0    0    0  ...    1    1    0   
1         0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
2         0    0    0    0    1    1    0    0    0    0  ...    0    0    0   
3         0    0    0    0    1    1    0    0    0    0  ...    0    0    0   
4         1    0    0    0    0    0    0    0    0    0  ...    0    0    0   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
804409    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
804410    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
804411    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
804412    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
804413    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   

        96   97   98   99   100  101  102  
0         0    0    0    0    0    0    1  
1         0    0    0    0    0    0    0  
2         0    0    0    0    0    0    0  
3         0    0    0    0    0    0    0  
4         0    0    0    0    0    0    0  
...     ...  ...  ...  ...  ...  ...  ...  
804409    0    0    0    0    0    0    0  
804410    0    0    0    0    0    0    0  
804411    0    0    0    0    0    0    0  
804412    0    0    0    0    0    0    0  
804413    0    0    0    0    0    0    0  

[804414 rows x 103 columns]

### Pick 1 from 103 classes
Classfiication problem: 
Decide whether a sample belongs to topic C11 or not 

In [6]:
rcv1.target_names[0]

'C11'

In [7]:
rcv1_target = pd.DataFrame.sparse.from_spmatrix(rcv1.target)[[0]]

In [8]:
rcv1_target.head()

0
0  0
1  0
2  0
3  0
4  1

In [9]:
# dependent variable
y_rcv1 = rcv1_target.rename(columns={0:'C11'})
# Independent variable
X_rcv1 = pd.DataFrame.sparse.from_spmatrix(rcv1.data)

In [10]:
y_rcv1.shape, X_rcv1.shape

((804414, 1), (804414, 47236))

In [11]:
y_rcv1.head()

C11
0    0
1    0
2    0
3    0
4    1

In [12]:
X_rcv1.head()

0      1      2      3      4      5      6      7      8      9      ...  \
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   

   47226  47227  47228  47229  47230  47231  47232  47233  47234  47235  
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 47236 columns]

### Data cleaning

In [13]:
y_rcv1.isna().sum()

C11    0
dtype: int64

In [14]:
X_rcv1.isna().sum().sum()

0

### Feature selection 

Select all samples(rows) that's been classified as C11

In [15]:
X_rcv1['C11'] = y_rcv1['C11']

In [16]:
X_rcv1_selected = X_rcv1[X_rcv1['C11']==1]

In [17]:
X_rcv1_selected.shape

(24325, 47237)

In [18]:
X_rcv1_selected.head()

0    1    2    3    4    5    6    7    8    9  ...  47227  47228  \
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
31  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
35  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
62  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   
71  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...    0.0    0.0   

    47229  47230  47231  47232  47233  47234  47235  C11  
4     0.0    0.0    0.0    0.0    0.0    0.0    0.0    1  
31    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1  
35    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1  
62    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1  
71    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1  

[5 rows x 47237 columns]

Count the number of appearance of each feature, then sort then in descending order

In [19]:
selected_features = X_rcv1_selected.sum().sort_values(ascending=False)[:11]

Pick the best 10 features, cause these 10 features are most likely to contribute to C11

In [20]:
selected_features

C11      24325.000000
8608       875.162896
44457      838.719646
20987      665.023277
32243      607.420891
20115      604.340757
6163       517.067412
16808      501.852491
30124      480.485297
33169      478.784971
31179      448.218424
dtype: float64

In [21]:
X_rcv1 = X_rcv1[selected_features.index[1:]]

In [22]:
X_rcv1 # indepent variables

8608     44457    20987     32243     20115      6163     16808  \
0       0.000000  0.000000  0.00000  0.000000  0.032871  0.000000  0.000000   
1       0.032901  0.000000  0.00000  0.020775  0.064230  0.000000  0.020642   
2       0.070132  0.059803  0.05151  0.000000  0.030506  0.000000  0.000000   
3       0.070132  0.059803  0.05151  0.000000  0.030506  0.000000  0.000000   
4       0.043679  0.000000  0.00000  0.000000  0.000000  0.000000  0.000000   
...          ...       ...      ...       ...       ...       ...       ...   
804409  0.030799  0.000000  0.00000  0.040813  0.000000  0.050839  0.000000   
804410  0.036575  0.000000  0.00000  0.000000  0.000000  0.000000  0.000000   
804411  0.024979  0.000000  0.00000  0.000000  0.000000  0.000000  0.000000   
804412  0.018623  0.000000  0.00000  0.024678  0.043851  0.000000  0.000000   
804413  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000  0.000000   

           30124     33169  31179  
0       0.000000  0.031187    0.0  
1       0.021806  0.049363    0.0  
2       0.000000  0.060741    0.0  
3       0.000000  0.060741    0.0  
4       0.000000  0.000000    0.0  
...          ...       ...    ...  
804409  0.000000  0.000000    0.0  
804410  0.000000  0.000000    0.0  
804411  0.000000  0.000000    0.0  
804412  0.000000  0.075669    0.0  
804413  0.032460  0.000000    0.0  

[804414 rows x 10 columns]

Remove all rows where all values are 0

In [23]:
y_rcv1 #dependent variables

C11
0         0
1         0
2         0
3         0
4         1
...     ...
804409    0
804410    0
804411    0
804412    0
804413    0

[804414 rows x 1 columns]

### Train/validation/test split

In [24]:
from sklearn.model_selection import train_test_split
import math

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X_rcv1, y_rcv1, test_size=0.3)

In [26]:
X_train.shape, X_test.shape

((563089, 10), (241325, 10))

In [27]:
X_train.head()

8608  44457  20987     32243     20115  6163     16808     30124  \
711417  0.000000    0.0    0.0  0.000000  0.000000   0.0  0.000000  0.000000   
573456  0.067815    0.0    0.0  0.000000  0.061195   0.0  0.057937  0.000000   
296537  0.000000    0.0    0.0  0.025563  0.000000   0.0  0.000000  0.026832   
218447  0.042132    0.0    0.0  0.097140  0.000000   0.0  0.000000  0.034612   
229663  0.000000    0.0    0.0  0.000000  0.000000   0.0  0.000000  0.000000   

           33169  31179  
711417  0.000000    0.0  
573456  0.000000    0.0  
296537  0.000000    0.0  
218447  0.032834    0.0  
229663  0.000000    0.0

In [28]:
y_train.head()

C11
711417    0
573456    0
296537    0
218447    0
229663    0

Further divide test data in 50:50 

In [29]:
X_vali = X_test[:math.floor(len(X_test)/2)]
X_test_new = X_test[math.floor(len(X_test)/2):]

In [30]:
X_vali.shape

(120662, 10)

In [31]:
X_test_new.shape

(120663, 10)

### Build model 

In [32]:
from sklearn import svm
from sklearn.linear_model import LogisticRegression
from scipy.sparse import csr_matrix

In [33]:
clf = svm.SVC()

In [34]:
# sparse matrix bug
X_train_new = np.asarray(csr_matrix(X_train.values).todense())

In [35]:
type(X_train_new)

numpy.ndarray

In [ ]:
clf.fit(X_train_new, y_train.values.flatten())
# too much time to, didn't get results yet

### Evaluation on validation set

In [ ]:
clf.predict(X_vali)